File: coding/H1_Graph_Connectivity_BFS/Solution/Solution.py
from collections import deque

n, m, q = map(int, input().split())

adj = [[] for _ in range(n + 1)]

for _ in range(m):
    u, v = map(int, input().split())
    adj[u].append(v)
    adj[v].append(u)

for _ in range(q):
    u, v = map(int, input().split())

    vis = [False] * (n + 1)
    dq = deque([u])
    vis[u] = True
    found = False

    while dq:
        x = dq.popleft()
        if x == v:
            found = True
            break
        for nb in adj[x]:
            if not vis[nb]:
                vis[nb] = True
                dq.append(nb)

    print("Yes" if found else "No")

File: coding/H1_Graph_Connectivity_BFS/Solution/Solution.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int m = sc.nextInt();
        int q = sc.nextInt();

        List<Integer>[] adj = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++)
            adj[i] = new ArrayList<>();

        // Read edges (undirected)
        for (int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            adj[u].add(v);
            adj[v].add(u);
        }

        while (q-- > 0) {
            int s = sc.nextInt();
            int t = sc.nextInt();

            boolean[] vis = new boolean[n + 1];
            Queue<Integer> queue = new LinkedList<>();

            queue.add(s);
            vis[s] = true;

            boolean found = false;

            while (!queue.isEmpty()) {
                int x = queue.poll();

                if (x == t) {
                    found = true;
                    break;
                }

                for (int nb : adj[x]) {
                    if (!vis[nb]) {
                        vis[nb] = true;
                        queue.add(nb);
                    }
                }
            }

            System.out.println(found ? "Yes" : "No");
        }

        sc.close();
    }
}
